### PR TITLE
test(core): add property-based fuzz testing framework using eris

### PR DESCRIPTION
**Pull request by RigaPaul**
### 1. Why is this change necessary?

Shopware has no property-based ("fuzz") testing today — only example-based tests, which only prove correctness for the specific inputs someone thought to write down. This PR introduces a minimal, production-grade property-based testing framework so edge cases that nobody thought to hand-write can be caught automatically.

### 2. What does this change do, exactly?

- Adds `giorgiosironi/eris` (a QuickCheck-style, PHPUnit-native property-based testing library) as a dev-only dependency. No runtime dependency, no new external tooling required in CI or locally beyond PHPUnit.
- Adds a new `fuzz` PHPUnit testsuite (`tests/fuzz/`, namespace `Shopware\Tests\Fuzz\`), run the same way as `unit`/`integration`/`migration` today.
- Registers `Shopware\Tests\Fuzz\` in the PHPStan `allowedUnitTestClassNamespaces` list, so fuzz tests are held to the same static-analysis bar as unit tests (mandatory `#[Package]`, `#[CoversClass]`).
- Adds four example fuzz tests demonstrating the pattern on real components:
  - `RangeFilterFuzzTest` — validation invariant on `RangeFilter` (invalid input is always rejected, never silently accepted).
  - `QueryStringParserFuzzTest` — round-trip invariant on `QueryStringParser` (serializing a parsed filter and re-parsing it never loses information, within its documented supported domain).
  - `HtmlSanitizerFuzzTest` — security invariant on `HtmlSanitizer` (sanitized output never contains a live `<script>` element, event-handler attribute, or `javascript:` URI), fuzzed against a corpus of OWASP-style XSS payloads combined with attribute-context breakout attempts.
  - `FieldSerializerFuzzTest` — round-trip and validation invariants on `FieldSerializer`'s ImportExport association parsing.
- Adds `tests/fuzz/README.md`: a plain-language explainer for engineers and QA who have no prior fuzz-testing background — what property-based testing is, why Shopware has it (using the two real bugs above as evidence), what kinds of functions/classes are good vs. poor fuzz targets, how Eris's shrinking and `ERIS_SEED` reproduction work, how to run the suite locally, and what to do when a test fails.

**Known, intentional exception:** `FieldSerializerFuzzTest::testRejectsASingleIdentifierContainingThePipeDelimiter` currently **fails**. It asserts the contract `FieldSerializer::normalizeId()`'s own dead guard implies (reject an identifier containing `"|"`), and was deliberately left failing rather than adjusted to pass, so it stays visible as a tracked, reproducible finding of the bug described above. It is not fixed as part of this PR — that's a separate decision about `FieldSerializer`'s intended behavior for pipe-delimited import data.

This PR does **not** include CI wiring (the fuzz suite currently runs locally only) or the `.agents/skills/shopware-fuzz-tests` authoring skill — both are natural follow-ups, not included here to keep this PR scoped to standing up the framework itself.

### 3. How to run the tests on the CLI

```bash
# run the whole fuzz suite
vendor/bin/phpunit --testsuite fuzz

# run a single component
vendor/bin/phpunit tests/fuzz/Core/Framework/DataAbstractionLayer/Search/Filter/RangeFilterTest.php

# reproduce the known FieldSerializer finding directly
vendor/bin/phpunit --testsuite fuzz --filter FieldSerializerTest

### CODE REVIEW

File: src/Core/DevOps/StaticAnalyze/PHPStan/common.neon [hunk 1]
```diff
@@ -68,6 +68,7 @@ parameters:
         allowedUnitTestClassNamespaces:
             - Shopware\Tests\Unit\
             - Shopware\Tests\Migration\
+            - Shopware\Tests\Fuzz\
```
Comment (nfortier-shopware) [inline src/Core/DevOps/StaticAnalyze/PHPStan/common.neon]: Shopware\Tests\PropertyBased please

We don't want to add more ambiguity (we already have a 'devops' suite which should be renamed 'staticAnalysis")

File: tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+ * can ever still contain a "|". The guard is unreachable. testRejectsASingleIdentifierContai
+ * ningThePipeDelimiter() asserts the contract that guard implies; it currently fails, which is
+ * the point - it demonstrates the bug rather than working around it. See
+ * .agents/skills/shopware-fuzz-tests for the pattern and .github/AGENTS.md-style "don't adjust
+ * the test to make it pass" discipline this file follows.
+ *
+ * @internal
+ */
+#[Package('fundamentals@after-sales')]
+#[CoversClass(FieldSerializer::class)]
```
Comment (nfortier-shopware) [inline tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php]: Let's not introduce a new coverage report, coverage metric belongs to unit.

Would you want it anyway, then it need to be added to the configuration of codecov lanes (we have already unit & migration)

File: tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+class FieldSerializerFuzzTest extends TestCase
```
Comment (nfortier-shopware) [inline tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php]: Please do not use `Fuzz` in class naming, the namespace is enough to distinguish from FieldSerializerTest of other suites

File: tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+{
+    use TestTrait;
```
Comment (nfortier-shopware) [inline tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php]: we are moving out from trait usage for various reason, let's create a class `FuzzTestCase` that all your tests classes will extend.

File: tests/fuzz/README.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+This directory holds Shopware's **property-based tests** ("fuzz tests"). This document explains
+what that means, why it exists, and how to run it — no prior fuzz-testing experience assumed.
+
+## What this is
+
+A normal unit test picks a handful of example inputs and checks the output for each one. A
+property-based test instead states a **rule that should always be true** ("an invariant") and
+then generates a large number of random inputs — around 100 per test by default — to try to
+break that rule.
```
Comment (nfortier-shopware) [inline tests/fuzz/README.md]: We will stick with property-based tests, "fuzz" is too vague

File: phpunit.xml.dist [hunk 6]
```diff
@@ -228,6 +228,10 @@
         <testsuite name="migration">
             <directory>tests/migration</directory>
         </testsuite>
+        <!-- Property-based tests (see .agents/skills/shopware-fuzz-tests); slow/noisy by nature, run nightly only -->
+        <testsuite name="fuzz">
+            <directory>tests/fuzz</directory>
```
Comment (nfortier-shopware) [inline phpunit.xml.dist]: name=property-based
tests/propertyBased

File: composer.json [hunk 7]
```diff
@@ -194,6 +194,7 @@
         "dominikb/composer-license-checker": "^2.5",
         "ergebnis/phpunit-slow-test-detector": "^2.4",
         "friendsofphp/php-cs-fixer": "3.95.15",
+        "giorgiosironi/eris": "^1.1",
```
Comment (nfortier-shopware) [inline composer.json:197]: This library is not compatible with PHPUnit 8.5 or 8.6 which could be a problem but we can try on the side

**Review by nfortier-shopware — CHANGES_REQUESTED** [review]
Nice initiative, but in the middle of migrating all our existing suite to PHPUnit 12/13 it would be great to attach it with a ticket for framework

All kind of experimentation as such, we could also go further like using `Infection` to mutate the code would need evaluation too, happy to help with that

File: tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            });
+    }
+
+    /**
+     * A single business-name-style value (e.g. a category or tag technical name a user typed,
+     * not a delimiter-separated list) that happens to contain a literal "|" must be rejected -
+     * per FieldSerializer::normalizeId()'s own str_contains($id, '|') guard - not silently
+     * split and misinterpreted as two associations to unrelated (and likely nonexistent)
+     * entities. This currently fails: see the class docblock.
+     */
```
Comment (nfortier-shopware) [inline tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php]: But the point is to not let pass failing tests no?

File: composer.json (hunk 7 above)
Comment (nfortier-shopware) [reply composer.json:197]: do not get blocked by this comment, though Shopware (including our other tests in PHPUnit suites) is already compatible with php 8.5, we still have work to do for 8.6. This is more a side note

File: tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+ * explode('|', ...) with no escaping, so a value that itself contains "|" cannot be
+ * distinguished from two values separated by "|".
+ *
+ * FieldSerializer::normalizeId() visibly intends to guard against this - it explicitly checks
+ * `str_contains($id, '|')` and throws InvalidIdentifierException - but deserialize() calls
+ * explode('|', $value) BEFORE normalizeId() ever sees an individual id, so no id it receives
+ * can ever still contain a "|". The guard is unreachable. testRejectsASingleIdentifierContai
+ * ningThePipeDelimiter() asserts the contract that guard implies; it currently fails, which is
+ * the point - it demonstrates the bug rather than working around it. See
+ * .agents/skills/shopware-fuzz-tests for the pattern and .github/AGENTS.md-style "don't adjust
```
Comment (nfortier-shopware) [inline tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php]: where is it? The agent file is missing no? 👀

File: tests/fuzz/Core/Framework/DataAbstractionLayer/Search/Parser/QueryStringParserFuzzTest.php [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    //             new SearchRequestException()
+    //         );
+    //
+    //         $reparsed = QueryStringParser::fromArray(
+    //             $definition,
+    //             QueryStringParser::toArray($parsed),
+    //             new SearchRequestException()
+    //         );
+    //
+    //         static::assertEquals($parsed, $reparsed);
```
Comment (nfortier-shopware) [inline tests/fuzz/Core/Framework/DataAbstractionLayer/Search/Parser/QueryStringParserFuzzTest.php]: Perhaps we could put 
```php
static::assertNotEquals($parsed, $reparsed, 'This is a bug, it should equals lalala');
```
Because then we would adapt it the moment we fix it, while commented test really nobody will truly consider it

File: tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+                    ));
+                } catch (InvalidIdentifierException) {
+                }
+            });
+    }
+
+    /**
+     * @phpstan-ignore missingType.generics (Eris's Generator only declares a Psalm template, not a PHPStan-compatible one)
+     */
+    private function safeName(): Generator
```
Comment (keulinho) [inline tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php]: i don't get the real benefit here 🤔 

We still need to pretty much hard code all the inputs, so what's the fuzzy way about it? 🤔 From the 3 examples in here, i guess all of them can become regular phpunit test cases with data providers that provide the different input options, to me that would be way easier to read and understand, those tests look unfamilliar and because of that i needed some time to understand what they are really testing

also i would be interested in how the issue for the query string parser was found? By fuzzing the input? Or did ai find and then created a fuzzy integration test as regression? 🤔

Comment (keulinho) [reply tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php]: ok i should have started with the readme first, that answers a lot of my questions 🤣 

Wondering where and how we can best use it, because especially the sanitizer test currently looks to me like we hard code all the cases 

But i see now the value for all parser type of classes

File: tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php (hunk 3 above)
Comment (RigaPaul) [reply tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php]: Renamed, but now if we want to run a specific test it can match multiple tests, eg.

`vendor/bin/phpunit --filter RangeFilterTest  // Matches both unit and fuzz tests`

So need to do:

`vendor/bin/phpunit --testsuite fuzz --filter RangeFilterTest`

File: tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php (hunk 11 above)
Comment (RigaPaul) [reply tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php]: The tests do look weird at first I have to admit. Because they are basically setting up the input range we want to test (eg. any string, number, specific tokens, etc...) and letting the runner compute the actual inputs within those ranges for each invocation.

The 2 issues were found by creating the tests - as soon as written the tests failed then the cause was analyzed.

File: composer.json (hunk 7 above)
Comment (RigaPaul) [reply composer.json:197]: No worries, thanks for flagging this. I will take a look, just to make sure that A) it's not a stale library and B) there isn't a more active alternative.

File: tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php (hunk 3 above)
Comment (nfortier-shopware) [reply tests/fuzz/Core/Content/ImportExport/DataAbstractionLayer/Serializer/Field/FieldSerializerFuzzTest.php]: Yes it is safer anyway to be explicit with the testsuite, if I recall well there is something about PHPUnit taking the whole tree of src/tests when they bubble a deprecation for instance so, for the PHPUnit runner it is better

File: composer.json (hunk 7 above)
Comment (RigaPaul) [reply composer.json:197]: Had a look - the lib is used quite a bit and isn't abandoned but is in 'conservative maintenance'. I will raise a PR to them for 8.5.
Unfortunately all other alternatives are worse.
